app/models/concerns/effective_devise_user.rb

```diff
@@ -162,6 +166,32 @@ def send_reset_password_instructions(attributes = {})
       recoverable
     end
 
+    # https://github.com/heartcombo/devise/blob/f6e73e5b5c8f519f4be29ac9069c6ed8a2343ce4/lib/devise/models/authenticatable.rb#L276
+    def find_first_by_auth_conditions(tainted_conditions, opts = {})
+      conditions = devise_parameter_filter.filter(tainted_conditions).merge(opts)
+      email = conditions[:email]
+      conditions.delete(:email)
+
+      user = to_adapter.find_first(conditions.merge(email: email))
+      return user if user.present? && user.persisted?
+
+      to_adapter.find_first(conditions.merge(alternate_email: email))
+    end
+
+    # https://github.com/heartcombo/devise/blob/f6e73e5b5c8f519f4be29ac9069c6ed8a2343ce4/lib/devise/models/database_authenticatable.rb#L216
+    def find_for_database_authentication(warden_conditions)
+      conditions = warden_conditions.dup.presence || {}
+      primary_or_alternate_email = conditions[:email]
+      conditions.delete(:email)
+
+      raise "Expected an email or alternate email but got [#{primary_or_alternate_email}] instead" if primary_or_alternate_email.blank?
+
+      all
+        .where(conditions)
+        .where("lower(email) = :value OR lower(alternate_email) = :value", value: primary_or_alternate_email.strip.downcase)
+        .take
```

Comment:
Interestingly, I think they might have changed this at some point because when I run them here, both use a `LIMIT 1` in the query. I believe that the only difference is that `.first` will add an `ORDER BY id ASC` which might only slow down some queries if there are too many records because the database has to sort before limiting to 1, right?

For `users` it won't make any perceivable difference in performance until we get to the hundreds of thousands of records, possibly? 🤔